include/dali/core/mm/detail/align.h

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_DETAIL_ALIGN_H_
+#define DALI_CORE_MM_DETAIL_ALIGN_H_
+
+#include <cassert>
+#include <utility>
+#include "dali/core/util.h"
+
+namespace dali {
+namespace mm {
+namespace detail {
+
+DALI_HOST_DEV
+constexpr uintptr_t ptr2u(void *p) noexcept {
+  // reinterpret_cast is not a constant expression - use a series of static casts instead
+  return static_cast<char*>(p) - static_cast<char*>(nullptr);
+}
+
+template <typename T>
+DALI_HOST_DEV
+constexpr T *u2ptr(uintptr_t i) noexcept {
+  // reinterpret_cast is not a constant expression - use a series of static casts instead
+  return static_cast<T*>(static_cast<void *>(static_cast<char*>(nullptr) + i));
+}
+
+template <typename T, typename A>
+DALI_HOST_DEV
+constexpr T *align_ptr(T *ptr, A alignment) noexcept {
+  return u2ptr<T>(align_up(ptr2u(ptr), alignment));
+}
+
+DALI_HOST_DEV
+constexpr bool is_aligned(void *ptr, size_t alignment) noexcept {
+  return (ptr2u(ptr) & (alignment-1)) == 0;
+}
+
+template <typename OffsetType, typename AllocFn>
+DALI_HOST_DEV
+void *aligned_alloc_impl(AllocFn base_alloc, size_t size, size_t alignment) {
+  void *ptr = base_alloc(size + alignment + sizeof(OffsetType) - 1);
```

Comment:
I'm probably missing something simple, why -1?

Reply:
Perhaps it should be ```size + std::max(alignment, sizeof(OffsetType))```. I'll give it a try (and run a test with alignment specifier above 256).

Reply:
In any case, what's there is more conservative than it needs to be.

Reply:
Hmm... maybe it was correct after all.

Reply:
It must be. The test checks it :)
Example:
```
alignment = 32
size = 1
OffsetType = uin64_t (8 bytes)
requested space is 40 bytes (1 + 32 + 8 - 1)

raw_pointer = 25  // worst case
buffer_start = align_up(25 + 8, 32) = align_up(33, 32) = 64
buffer_end: 64+1 (suable size is 1) = 65
size_used = buffer_end - raw_pointer = 65-25 = 40 -> that's as much as we requested

now, let's see the neighboring cases (raw pointer 24 and 26):
raw_pointer = 24
buffer_start = align_up(24 + 8 ,32) = align_up(32, 32) = 32
buffer_end = 32 + 1 = 33
size_used = buffer_end - raw_pointer = 33 - 24 = 9  // much less than requested

raw_pointer = 26
buffer_start = align_up(26 + 8, 32) = align_up(34, 32) = 64
buffer_end: 64+1 (suable size is 1) = 65
size_used = buffer_end - raw_pointer = 65-26 = 39  // less than the pessimistic estimate
```

Reply:
Now i get it. Thanks for the explanation.